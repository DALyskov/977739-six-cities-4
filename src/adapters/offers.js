export const createOffers = (data) =>
  data.map((offer) => {
    return {
      bedrooms: offer.bedrooms,
      city: offer.city,
      description: offer.description,
      features: offer.goods,

      hostName: offer.host.name,
      hostAvatar: offer.host.avatar_url,
      isHostPro: offer.host.is_pro,
      hostId: offer.host.id,

      id: offer.id,
      images: offer.images,
      isBookmark: offer.is_favorite,
      isPremium: offer.is_premium,

      location: offer.location,

      maxAdults: offer.max_adults,
      previewImg: offer.preview_image,
      price: offer.price,
      rating: offer.rating,
      name: offer.title,
      type: offer.type,
    };
  });
