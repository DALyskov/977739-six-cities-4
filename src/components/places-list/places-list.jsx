import React, {PureComponent} from 'react';
import propTypes from 'prop-types';
import {connect} from 'react-redux';

import {
  PlacesClassNames,
  sortingItems,
  SortingTypeDict,
  PageType,
} from '../../const.js';

import {Operation as DataOperation} from '../../reducer/data/data.js';
import {ActionCreator as AppActionCreator} from '../../reducer/state-application/state-application.js';
import {getSortingType} from '../../reducer/state-application/selectors.js';
import PlaceCard from '../place-card/place-card.jsx';

class PlaceList extends PureComponent {
  _sortOffersBytype(offers, sortingType) {
    let sortedOffers = offers.slice();
    switch (sortingType) {
      case sortingItems[1]:
        return sortedOffers.sort(
          (a, b) => a[SortingTypeDict.PRICE] - b[SortingTypeDict.PRICE]
        );
      case sortingItems[2]:
        return sortedOffers.sort(
          (a, b) => b[SortingTypeDict.PRICE] - a[SortingTypeDict.PRICE]
        );
      case sortingItems[3]:
        return sortedOffers.sort(
          (a, b) => b[SortingTypeDict.RATING] - a[SortingTypeDict.RATING]
        );
      default:
        return offers;
    }
  }

  render() {
    const {
      offersByCity,
      className,
      sortingType,
      onPlaceCardNameClick,
      onPlaceCardHover,
      onFavoriteBtnClick,
    } = this.props;

    const sortedOffers = this._sortOffersBytype(offersByCity, sortingType);

    return (
      <div className={`${className[0]} places__list tabs__content`}>
        {sortedOffers.map((placeData) => (
          <PlaceCard
            key={placeData.id}
            placeData={placeData}
            className={className[1]}
            onPlaceCardNameClick={onPlaceCardNameClick}
            onPlaceCardHover={onPlaceCardHover}
            onFavoriteBtnClick={onFavoriteBtnClick}
          />
        ))}
      </div>
    );
  }
}

PlaceList.propTypes = {
  offersByCity: propTypes.arrayOf(
    propTypes.shape({
      bedrooms: propTypes.number.isRequired,
      city: propTypes.shape({
        location: propTypes.shape({
          latitude: propTypes.number.isRequired,
          longitude: propTypes.number.isRequired,
          zoom: propTypes.number.isRequired,
        }).isRequired,
        name: propTypes.string.isRequired,
      }),
      description: propTypes.string.isRequired,
      features: propTypes.arrayOf(propTypes.string.isRequired),
      hostName: propTypes.string.isRequired,
      hostAvatar: propTypes.string.isRequired,
      isHostPro: propTypes.bool,
      hostId: propTypes.number.isRequired,
      id: propTypes.number.isRequired,
      images: propTypes.arrayOf(propTypes.string.isRequired).isRequired,
      isBookmark: propTypes.bool,
      isPremium: propTypes.bool,
      location: propTypes.shape({
        latitude: propTypes.number.isRequired,
        longitude: propTypes.number.isRequired,
        zoom: propTypes.number.isRequired,
      }).isRequired,
      maxAdults: propTypes.number.isRequired,
      previewImg: propTypes.string.isRequired,
      price: propTypes.number.isRequired,
      rating: propTypes.number.isRequired,
      name: propTypes.string.isRequired,
      type: propTypes.string.isRequired,
    })
  ).isRequired,

  className: propTypes.oneOf(Object.values(PlacesClassNames)).isRequired,

  sortingType: propTypes.oneOf(sortingItems).isRequired,

  onPlaceCardNameClick: propTypes.func.isRequired,
  onPlaceCardHover: propTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
  sortingType: getSortingType(state),
});

const mapDispatchToProps = (dispatch) => ({
  onPlaceCardNameClick(placeData) {
    dispatch(AppActionCreator.changePlace(placeData));
    dispatch(AppActionCreator.changeActivePage(PageType.PROPERTY));
    dispatch(DataOperation.loadReviews(placeData.id));
    dispatch(DataOperation.loadNearbyOffers(placeData.id));
  },
  onPlaceCardHover(placeDataId) {
    dispatch(AppActionCreator.changeHoverCityId(placeDataId));
  },
  onFavoriteBtnClick(placeDataId, placeDataIsBookmark) {
    dispatch(DataOperation.sendFavoriteOffer(placeDataId, placeDataIsBookmark));
  },
});

export {PlaceList};
export default connect(mapStateToProps, mapDispatchToProps)(PlaceList);
