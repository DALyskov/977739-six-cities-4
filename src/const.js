export const APPROVED_NAME = [
  `Beautiful & luxurious apartment at great location`,
  `Wood and stone place`,
  `Canal View Prinsengracht`,
  `Nice, cozy, warm big bed apartment`,
];
