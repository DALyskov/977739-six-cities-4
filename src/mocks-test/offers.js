export const offers = [
  {
    bedrooms: 1,
    city: {
      location: {
        latitude: 52.37454,
        longitude: 4.897976,
        zoom: 13,
      },
      name: `Amsterdam`,
    },
    description: `I rent out a very sunny and bright apartment only 7 minutes walking distance to the metro station. The apartment has a spacious living room with a kitchen, one bedroom and a bathroom with mit bath. A terrace can be used in summer.`,
    features: [
      `Coffee machine`,
      `Laptop friendly workspace`,
      `Air conditioning`,
      `Baby seat`,
      `Fridge`,
      `Breakfast`,
      `Dishwasher`,
      `Towels`,
      `Washer`,
      `Washing machine`,
    ],

    hostName: `Angelina`,
    hostAvatar: `img/avatar-angelina.jpg`,
    isHostPro: true,
    hostId: 25,

    id: 5,
    images: [
      `img/apartment-01.jpg`,
      `img/apartment-02.jpg`,
      `img/apartment-03.jpg`,
    ],
    isBookmark: false,
    isPremium: false,

    location: {
      latitude: 52.397540000000006,
      longitude: 4.9099759999999995,
      zoom: 16,
    },

    maxAdults: 2,
    previewImg: `img/room.jpg`,
    price: 101,
    rating: 3.5,
    name: `Perfectly located Castro`,
    type: `room`,
  },

  {
    bedrooms: 5,
    city: {
      location: {
        latitude: 52.37454,
        longitude: 4.897976,
        zoom: 13,
      },
      name: `Amsterdam`,
    },
    description: `I rent out a very sunny and bright apartment only 7 minutes walking distance to the metro station. The apartment has a spacious living room with a kitchen, one bedroom and a bathroom with mit bath. A terrace can be used in summer.`,
    features: [
      `Coffee machine`,
      `Laptop friendly workspace`,
      `Air conditioning`,
      `Baby seat`,
      `Fridge`,
      `Breakfast`,
      `Dishwasher`,
      `Towels`,
      `Washer`,
      `Washing machine`,
    ],

    hostName: `Angelina`,
    hostAvatar: `img/avatar-angelina.jpg`,
    isHostPro: true,
    hostId: 25,

    id: 1,
    images: [`img/apartment-01.jpg`, `img/apartment-02.jpg`],
    isBookmark: false,
    isPremium: false,

    location: {
      latitude: 52.397540000000006,
      longitude: 4.886976,
      zoom: 16,
    },

    maxAdults: 6,
    previewImg: `img/room.jpg`,
    price: 813,
    rating: 2.4,
    name: `Wood and stone place`,
    type: `house`,
  },

  {
    bedrooms: 4,
    city: {
      location: {
        latitude: 52.37454,
        longitude: 4.897976,
        zoom: 13,
      },
      name: `Amsterdam`,
    },
    description: `This is a place for dreamers to reset, reflect, and create. Designed with a 'slow' pace in mind, our hope is that you enjoy every part of your stay; from making local coffee by drip in the morning, choosing the perfect record to put on as the sun sets.`,
    features: [`Breakfast`, `Laptop friendly workspace`],

    hostName: `Angelina`,
    hostAvatar: `img/avatar-angelina.jpg`,
    isHostPro: true,
    hostId: 25,

    id: 14,
    images: [`img/apartment-01.jpg`, `img/apartment-03.jpg`],
    isBookmark: false,
    isPremium: true,

    location: {
      latitude: 52.397540000000006,
      longitude: 4.886976,
      zoom: 16,
    },

    maxAdults: 7,
    previewImg: `img/room.jpg`,
    price: 423,
    rating: 4.7,
    name: `Penthouse, 4-5 rooms + 5 balconies`,
    type: `hotel`,
  },
];

export const originOffers = [
  {
    bedrooms: 1,
    city: {
      location: {
        latitude: 52.37454,
        longitude: 4.897976,
        zoom: 13,
      },
      name: `Amsterdam`,
    },
    description: `I rent out a very sunny and bright apartment only 7 minutes walking distance to the metro station. The apartment has a spacious living room with a kitchen, one bedroom and a bathroom with mit bath. A terrace can be used in summer.`,
    goods: [
      `Coffee machine`,
      `Laptop friendly workspace`,
      `Air conditioning`,
      `Baby seat`,
      `Fridge`,
      `Breakfast`,
      `Dishwasher`,
      `Towels`,
      `Washer`,
      `Washing machine`,
    ],

    host: {
      'avatar_url': 'img/avatar-angelina.jpg',
      id: 25,
      'is_pro': true,
      name: 'Angelina',
    },
    id: 5,
    images: [
      `img/apartment-01.jpg`,
      `img/apartment-02.jpg`,
      `img/apartment-03.jpg`,
    ],
    'is_favorite': false,
    'is_premium': false,
    location: {
      latitude: 52.397540000000006,
      longitude: 4.9099759999999995,
      zoom: 16,
    },
    'max_adults': 2,
    'preview_image': `img/room.jpg`,
    price: 101,
    rating: 3.5,
    title: `Perfectly located Castro`,
    type: `room`,
  },
];
