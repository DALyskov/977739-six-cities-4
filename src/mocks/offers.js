export default [
  {
    id: 1,
    isPremium: true,
    images: [
      `img/apartment-01.jpg`,
      `img/room.jpg`,
      `img/apartment-02.jpg`,
      `img/apartment-03.jpg`,
    ],
    price: 120,
    isBookmark: false,
    rating: 4.5,
    name: `Beautiful & luxurious apartment at great location`,
    type: `Apartment`,
    bedrooms: 3,
    maxAdults: 1,
    features: [
      'Heating',
      'Kitchen',
      'Cable TV',
      'Washing machine',
      'Coffee machine',
    ],
    hostName: `Dima`,
    hostAvatar: `img/avatar-angelina.jpg`,
    isHostPro: true,
    description:
      'A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.',
    city: {
      location: {
        latitude: 52.370216,
        longitude: 4.895168,
        zoom: 10,
      },
      name: 'Amsterdam',
    },
    location: {
      latitude: 52.3909553943508,
      longitude: 4.85309666406198,
      zoom: 8,
    },
  },
  {
    id: 2,
    isPremium: false,
    images: [`img/room.jpg`],
    price: 80,
    isBookmark: true,
    rating: 4,
    name: `Wood and stone place`,
    type: `Private room`,
    bedrooms: 2,
    maxAdults: 3,
    features: [
      'Heating',
      'Kitchen',
      'Cable TV',
      'Washing machine',
      'Coffee machine',
      'Dishwasher',
    ],
    hostName: `Alex`,
    hostAvatar: `img/avatar-angelina.jpg`,
    isHostPro: false,
    description:
      'A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.',
    city: {
      location: {
        latitude: 52.370216,
        longitude: 4.895168,
        zoom: 10,
      },
      name: 'Amsterdam',
    },
    location: {
      latitude: 52.369553943508,
      longitude: 4.85309666406198,
      zoom: 8,
    },
  },
  {
    id: 3,
    isPremium: false,
    images: [`img/apartment-02.jpg`],
    price: 132,
    isBookmark: false,
    rating: 4,
    name: `Canal View Prinsengracht`,
    type: `Apartment`,
    bedrooms: 5,
    maxAdults: 2,
    features: ['Heating', 'Cable TV', 'Coffee machine', 'Dishwasher'],
    hostName: `Pavel`,
    hostAvatar: `img/avatar-angelina.jpg`,
    isHostPro: false,
    description:
      'A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.',
    city: {
      location: {
        latitude: 52.370216,
        longitude: 4.895168,
        zoom: 10,
      },
      name: 'Amsterdam',
    },
    location: {
      latitude: 52.3909553943508,
      longitude: 4.929309666406198,
      zoom: 8,
    },
  },
  {
    id: 5,
    isPremium: true,
    images: [
      `img/apartment-01.jpg`,
      `img/room.jpg`,
      `img/apartment-02.jpg`,
      `img/apartment-03.jpg`,
    ],
    price: 120,
    isBookmark: false,
    rating: 4.5,
    name: `Beautiful & luxurious apartment at great location`,
    type: `Apartment`,
    bedrooms: 3,
    maxAdults: 1,
    features: [
      'Heating',
      'Kitchen',
      'Cable TV',
      'Washing machine',
      'Coffee machine',
    ],
    hostName: `Dima`,
    hostAvatar: `img/avatar-angelina.jpg`,
    isHostPro: true,
    description:
      'A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.',
    city: {
      location: {
        latitude: 48.8652777777778,
        longitude: 2.3486111111111,
        zoom: 10,
      },
      name: 'Paris',
    },
    location: {
      latitude: 48.8653777777778,
      longitude: 2.3486111111111,
      zoom: 8,
    },
  },
  {
    id: 4,
    isPremium: true,
    images: [`img/apartment-03.jpg`],
    price: 180,
    isBookmark: false,
    rating: 5,
    name: `Nice, cozy, warm big bed apartment`,
    type: `Apartment`,
    bedrooms: 1,
    maxAdults: 4,
    features: ['Kitchen', 'Washing machine', 'Coffee machine', 'Dishwasher'],
    hostName: `Angelina`,
    hostAvatar: `img/avatar-angelina.jpg`,
    isHostPro: true,
    description:
      'A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.',
    city: {
      location: {
        latitude: 52.370216,
        longitude: 4.895168,
        zoom: 10,
      },
      name: 'Amsterdam',
    },
    location: {
      latitude: 52.3809553943508,
      longitude: 4.939309666406198,
      zoom: 8,
    },
  },
];
