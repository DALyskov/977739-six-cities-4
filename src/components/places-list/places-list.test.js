import React from 'react';
import renderer from 'react-test-renderer';

import {PlacesClassNames, sortingItems} from '../../const.js';
import {offers} from '../../mocks-test/offers.js';

import {PlaceList} from './places-list.jsx';

describe(`PlaceList_snapchots`, () => {
  it(`with_data`, () => {
    const tree = renderer
      .create(
        <PlaceList
          offersByCity={offers}
          className={PlacesClassNames.MAIN}
          sortingType={sortingItems[0]}
          onPlaceCardNameClick={() => {}}
          onPlaceCardHover={() => {}}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it(`without_data`, () => {
    const tree = renderer
      .create(
        <PlaceList
          offersByCity={[]}
          className={PlacesClassNames.MAIN}
          sortingType={sortingItems[0]}
          onPlaceCardNameClick={() => {}}
          onPlaceCardHover={() => {}}
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
