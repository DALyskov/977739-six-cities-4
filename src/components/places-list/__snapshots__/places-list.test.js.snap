// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaceList_snapchots with_data 1`] = `
<div
  className="cities__places-list places__list tabs__content"
>
  <article
    className="cities__place-card place-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    
    <div
      className="cities__image-wrapper place-card__image-wrapper"
    >
      <a
        href="#"
        onClick={[Function]}
      >
        <img
          alt="Place image"
          className="place-card__image"
          height="200"
          src="img/apartment-01.jpg"
          width="260"
        />
      </a>
    </div>
    <div
      className="place-card__info"
    >
      <div
        className="place-card__price-wrapper"
      >
        <div
          className="place-card__price"
        >
          <b
            className="place-card__price-value"
          >
            €
            101
          </b>
          <span
            className="place-card__price-text"
          >
            / night
          </span>
        </div>
        <button
          className="place-card__bookmark-button button"
          type="button"
        >
          <svg
            className="place-card__bookmark-icon"
            height="19"
            width="18"
          >
            <use
              xlinkHref="#icon-bookmark"
            />
          </svg>
          <span
            className="visually-hidden"
          >
            To bookmarks
          </span>
        </button>
      </div>
      <div
        className="place-card__rating rating"
      >
        <div
          className="place-card__stars rating__stars"
        >
          <span
            style={
              Object {
                "width": "80%",
              }
            }
          />
          <span
            className="visually-hidden"
          >
            Rating
          </span>
        </div>
      </div>
      <h2
        className="place-card__name"
        onClick={[Function]}
      >
        <a
          data-id={5}
          href="#"
        >
          Perfectly located Castro
        </a>
      </h2>
      <p
        className="place-card__type"
      >
        room
      </p>
    </div>
  </article>
  <article
    className="cities__place-card place-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    
    <div
      className="cities__image-wrapper place-card__image-wrapper"
    >
      <a
        href="#"
        onClick={[Function]}
      >
        <img
          alt="Place image"
          className="place-card__image"
          height="200"
          src="img/apartment-01.jpg"
          width="260"
        />
      </a>
    </div>
    <div
      className="place-card__info"
    >
      <div
        className="place-card__price-wrapper"
      >
        <div
          className="place-card__price"
        >
          <b
            className="place-card__price-value"
          >
            €
            813
          </b>
          <span
            className="place-card__price-text"
          >
            / night
          </span>
        </div>
        <button
          className="place-card__bookmark-button button"
          type="button"
        >
          <svg
            className="place-card__bookmark-icon"
            height="19"
            width="18"
          >
            <use
              xlinkHref="#icon-bookmark"
            />
          </svg>
          <span
            className="visually-hidden"
          >
            To bookmarks
          </span>
        </button>
      </div>
      <div
        className="place-card__rating rating"
      >
        <div
          className="place-card__stars rating__stars"
        >
          <span
            style={
              Object {
                "width": "40%",
              }
            }
          />
          <span
            className="visually-hidden"
          >
            Rating
          </span>
        </div>
      </div>
      <h2
        className="place-card__name"
        onClick={[Function]}
      >
        <a
          data-id={1}
          href="#"
        >
          Wood and stone place
        </a>
      </h2>
      <p
        className="place-card__type"
      >
        house
      </p>
    </div>
  </article>
  <article
    className="cities__place-card place-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="place-card__mark"
    >
      <span>
        Premium
      </span>
    </div>
    <div
      className="cities__image-wrapper place-card__image-wrapper"
    >
      <a
        href="#"
        onClick={[Function]}
      >
        <img
          alt="Place image"
          className="place-card__image"
          height="200"
          src="img/apartment-01.jpg"
          width="260"
        />
      </a>
    </div>
    <div
      className="place-card__info"
    >
      <div
        className="place-card__price-wrapper"
      >
        <div
          className="place-card__price"
        >
          <b
            className="place-card__price-value"
          >
            €
            423
          </b>
          <span
            className="place-card__price-text"
          >
            / night
          </span>
        </div>
        <button
          className="place-card__bookmark-button button"
          type="button"
        >
          <svg
            className="place-card__bookmark-icon"
            height="19"
            width="18"
          >
            <use
              xlinkHref="#icon-bookmark"
            />
          </svg>
          <span
            className="visually-hidden"
          >
            To bookmarks
          </span>
        </button>
      </div>
      <div
        className="place-card__rating rating"
      >
        <div
          className="place-card__stars rating__stars"
        >
          <span
            style={
              Object {
                "width": "100%",
              }
            }
          />
          <span
            className="visually-hidden"
          >
            Rating
          </span>
        </div>
      </div>
      <h2
        className="place-card__name"
        onClick={[Function]}
      >
        <a
          data-id={14}
          href="#"
        >
          Penthouse, 4-5 rooms + 5 balconies
        </a>
      </h2>
      <p
        className="place-card__type"
      >
        hotel
      </p>
    </div>
  </article>
</div>
`;

exports[`PlaceList_snapchots without_data 1`] = `
<div
  className="cities__places-list places__list tabs__content"
/>
`;
